metrics/metrics.go

```diff
@@ -101,26 +208,105 @@ var _ CSIMetricsManager = &csiMetricsManager{}
 
 type csiMetricsManager struct {
 	registry                   metrics.KubeRegistry
+	subsystem                  string
+	stabilityLevel             metrics.StabilityLevel
 	driverName                 string
-	csiOperationsMetric        *metrics.CounterVec
+	additionalLabelNames       []string
+	additionalLabels           []label
 	csiOperationsLatencyMetric *metrics.HistogramVec
 }
 
+type label struct {
+	name, value string
+}
+
 func (cmm *csiMetricsManager) GetRegistry() metrics.KubeRegistry {
 	return cmm.registry
 }
 
-// RecordMetrics must be called upon CSI Operation completion to record
-// the operation's metric.
-// operationName - Name of the CSI operation.
-// operationErr - Error, if any, that resulted from execution of operation.
-// operationDuration - time it took for the operation to complete
+// RecordMetrics implements CSIMetricsManager.RecordMetrics.
 func (cmm *csiMetricsManager) RecordMetrics(
 	operationName string,
 	operationErr error,
 	operationDuration time.Duration) {
-	cmm.csiOperationsLatencyMetric.WithLabelValues(
-		cmm.driverName, operationName, getErrorCode(operationErr)).Observe(operationDuration.Seconds())
+	cmm.recordMetricsWithLabels(operationName, operationErr, operationDuration)
+}
+
+// recordMetricsWithLabels is the internal implementation of RecordMetrics.
+func (cmm *csiMetricsManager) recordMetricsWithLabels(
+	operationName string,
+	operationErr error,
+	operationDuration time.Duration,
+	labelValues ...string) {
+	values := []string{cmm.driverName, operationName, getErrorCode(operationErr)}
+	toAdd := len(labelValues)
+	if toAdd > len(cmm.additionalLabelNames) {
+		// To many labels?! Truncate. Shouldn't happen because of
+		// error checking in WithLabelValues.
+		toAdd = len(cmm.additionalLabelNames)
+	}
+	values = append(values, labelValues[0:toAdd]...)
+	for i := toAdd; i < len(cmm.additionalLabelNames); i++ {
+		// Backfill missing values with empty string.
```

Comment:
I'm not sure if we can make the assumption that the missing labels are all at the end.